sirepo/auth/ldap.py

```diff
@@ -0,0 +1,56 @@
+"""LDAP login
+:copyright: Copyright (c) 2018-2022 RadiaSoft LLC.  All Rights Reserved.
+:license: http://www.apache.org/licenses/LICENSE-2.0.html
+"""
+from pykern import pkconfig
+from pykern.pkdebug import pkdlog, pkdp
+from pykern.pkcollections import PKDict
+import ldap
+import sirepo.quest
+
+AUTH_METHOD = "ldap"
+
+AUTH_METHOD_VISIBLE = True
+
+_cfg = None
+
+class API(sirepo.quest.API):
+    @sirepo.quest.Spec("require_cookie_sentinel")
+    def api_authLdapAuthorized(self):
+
+        # TODO(BG): pass simulation type as method parameter to assign to 'type', currently passing simulation_type to req and leaving 'type' False
+        req = self.parse_post(type=False)
+        u = req.req_data.username
+        p = req.req_data.password
+
+        c = ldap.initialize(_cfg.ldap_server)
+
+        if(len(u) > 256 or len(p) > 256):
+            raise Exception("login user/pass greater than 256 chars")
+
+        c.simple_bind_s(self._get_escaped(u), p)
+        
+        raise sirepo.util.Redirect(sirepo.uri.local_route(req.req_data.simulation_type))
```

Comment:
This seems incorrect. It probably needs to be a call to auth.login as auth.email does.

Reply:
Changing to auth.login, currently always registers a new user.

Want to verify I understand, to implement this properly I need to first search auth_db for a user "model" (by LDAP email?) to pass to auth.login, logging in with that user or registering a new user if None?

Reply:
This is a bit tricky. We do not necessarily want yet another table for users (auth_ldap_user_t), but I see no way around this. Perhaps it's ok, but there's code [that assumes METHOD_EMAIL](https://github.com/radiasoft/sirepo/blob/219a97bed9b610b27b12fc67a1058b3a0f3ef7e3/sirepo/auth_role_moderation.py#L82), which is (maybe) problematic. ldap could reuse email's model, because [there's other code](https://github.com/radiasoft/sirepo/blob/219a97bed9b610b27b12fc67a1058b3a0f3ef7e3/sirepo/pkcli/jupyterhublogin.py#L35) which relies on AuthEmailUser having emails. It's going to be tricky, but you could set ldap.UserModel = "AuthEmailUser".

I think you should authenticate with ldap to avoid the case where the user types the wrong email address. Once authenticated, you can then check to see if there's a user in auth_email_user_t. If there is, then you can use that uid to login (like api_authEmailLogin does).

Try this first.